<!DOCTYPE html>
<html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8"/>

<title>Introduction</title>

<script type="text/javascript">
window.onload = function() {
  var imgs = document.getElementsByTagName('img'), i, img;
  for (i = 0; i < imgs.length; i++) {
    img = imgs[i];
    // center an image if it is the only element of its parent
    if (img.parentElement.childElementCount === 1)
      img.parentElement.style.textAlign = 'center';
  }
};
</script>



<!-- MathJax scripts -->
<script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML">
</script>


<style type="text/css">
body, td {
   font-family: sans-serif;
   background-color: white;
   font-size: 13px;
}

body {
  max-width: 800px;
  margin: auto;
  padding: 1em;
  line-height: 20px;
}

tt, code, pre {
   font-family: 'DejaVu Sans Mono', 'Droid Sans Mono', 'Lucida Console', Consolas, Monaco, monospace;
}

h1 {
   font-size:2.2em;
}

h2 {
   font-size:1.8em;
}

h3 {
   font-size:1.4em;
}

h4 {
   font-size:1.0em;
}

h5 {
   font-size:0.9em;
}

h6 {
   font-size:0.8em;
}

a:visited {
   color: rgb(50%, 0%, 50%);
}

pre, img {
  max-width: 100%;
}
pre {
  overflow-x: auto;
}
pre code {
   display: block; padding: 0.5em;
}

code {
  font-size: 92%;
  border: 1px solid #ccc;
}

code[class] {
  background-color: #F8F8F8;
}

table, td, th {
  border: none;
}

blockquote {
   color:#666666;
   margin:0;
   padding-left: 1em;
   border-left: 0.5em #EEE solid;
}

hr {
   height: 0px;
   border-bottom: none;
   border-top-width: thin;
   border-top-style: dotted;
   border-top-color: #999999;
}

@media print {
   * {
      background: transparent !important;
      color: black !important;
      filter:none !important;
      -ms-filter: none !important;
   }

   body {
      font-size:12pt;
      max-width:100%;
   }

   a, a:visited {
      text-decoration: underline;
   }

   hr {
      visibility: hidden;
      page-break-before: always;
   }

   pre, blockquote {
      padding-right: 1em;
      page-break-inside: avoid;
   }

   tr, img {
      page-break-inside: avoid;
   }

   img {
      max-width: 100% !important;
   }

   @page :left {
      margin: 15mm 20mm 15mm 10mm;
   }

   @page :right {
      margin: 15mm 10mm 15mm 20mm;
   }

   p, h2, h3 {
      orphans: 3; widows: 3;
   }

   h2, h3 {
      page-break-after: avoid;
   }
}
</style>



</head>

<body>
<h1>Introduction</h1>

<p>The genetic information collected in genome-wide association studies (GWAS)
is represented by the genotypes of various single-nucleotid polymorphisms (SNPS).
Testing biological meaningful SNP Sets is a successful strategy for the
evaluation of GWAS data, as it may increase power as well as interpretation of
results. Via mapping of SNPs to genes forming a network, association between 
pathways and disease risk can be investigated.<br/>
Kernel methods are particularly well suited to cope with the challenges connected 
to the analysis of large SNP sets from GWAS data. They do not require to model
a direct functional relationship between SNPs and effects, while at the same 
time can deal with high-dimensional data and allow for straightforward
incorporation of covariates. The model for a logistic kernel machine regression 
of a pathway on a binary outcome is given by</p>

<p>\[logit(P(y_i=1|x_i,z_i))=x_i^t \beta + h(z_i)                    (1)        \]                             </p>

<p>where \(y_i\) denotes the case or control status of individual i, \(x_i\) is 
the vector including informative covariates (such as age, sex, etc.) and \(z_i\) 
represents the genotypes of individual i. \(\beta\) is the regression coefficients 
for the parametric part of the model, while \(h(.)\) denotes an unknown function, 
non-parametrically incorporating the pathway&#39;s influence. The intercept is 
assumed to be included in \(x_i\). For more details see Liu et al (2008).</p>

<p>Different kernels have been proposed that convert the genomic information of two
individuals into a quantitative value reflecting their genetic similarity. 
This package includes the linear kernel as well as two more advanced kernels, 
adjusting for size bias in the number of SNPs and genes in a pathway or 
incorporating the network structure of genes within the pathway, respectively. 
The kernel functions are described in more detail in the instructions below.</p>

<p>A variance component test, constructed around the similarity matrix, can be used 
to evaluate a pathway&#39;s influence on disease risk. In <code>kangar00</code> p-values can be
calculated with the Satterthwaite approximation or Davies method as described 
in Schaid (2010) and Davies (1980), respectively.</p>

<h1>Data extraction and preparation</h1>

<h2>Pathways</h2>

<p>The kangar00 package offers several functions for data extraction from internet 
databases. In the following they will be explained using the Circadian rhythm 
pathway as an example.</p>

<ul>
<li><p>In the KEGG database (Kanehisa et al 2014) this pathway is identified with 
the id <em>hsa04710</em>. </p></li>
<li><p>The function <code>pathway_info()</code> can use this id to create a table listing all 
genes included in Circadian rhythm. For each gene the startpoint, endpoint 
and the chromosome are listed.</p></li>
<li><p>Gene membership is obtained directly from KEGG, while startpoints, endpoints 
and chromosome information is extracted from Ensembl (Cunningham et al 2015).
The database is accessed via the function <code>getBM()</code> in the biomaRt package. 
This means that the gene boundaries given will equal the current build used 
in Ensemble. An internet connection is required for this step.   </p>

<pre><code>pathway_info(&#39;hsa04710&#39;)   
</code></pre>

<p>will return a <code>pathway_info</code> object containing a data frame of the form </p></li>
</ul>

<table><thead>
<tr>
<th align="center">pathway</th>
<th align="right">gene_start</th>
<th align="right">gene_end</th>
<th align="right">chr</th>
<th align="center">gene</th>
</tr>
</thead><tbody>
<tr>
<td align="center">hsa04710</td>
<td align="right">13276652</td>
<td align="right">13387266</td>
<td align="right">11</td>
<td align="center">ARNTL</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">4979116</td>
<td align="right">4985323</td>
<td align="right">3</td>
<td align="center">BHLHE40</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">26120026</td>
<td align="right">26125127</td>
<td align="right">12</td>
<td align="center">BHLHE41</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">&hellip;</td>
<td align="right">&hellip;</td>
<td align="right">&hellip;</td>
<td align="center">&hellip;</td>
</tr>
</tbody></table>

<p>listing information on all genes KEGG assigned to Circadian rhythm.</p>

<h3>Pathway object</h3>

<p>In <code>kangar00</code> all information on a specific pathway is combined in a <code>pathway</code>
object.  It includes </p>

<ul>
<li><p>The pathway&#39;s ID as used in KEGG. </p></li>
<li><p>The adjacency <code>matrix</code>, which equals the network matrix without signs. </p></li>
<li><p>A <code>vector</code> giving the signs for the interactions.</p></li>
</ul>

<p>The following example creates a new pathway object, to which gene-interaction
information has yet to be added</p>

<pre><code>pathw &lt;- pathway(id=&#39;hsa04710&#39;, adj=matrix(0), sign=as.vector(matrix(0)[matrix(0)!=0]))
</code></pre>

<h3>Networkmatrix</h3>

<p>The gene-gene interactions within pathways are represented by a network matrix. 
This quadratic matrix is of dimension equal to the number of genes in 
the corresponding pathway. It includes entries equal to \(1\) (representing an 
activation interaction), \(-1\) (denoting an inhibiting interaction) or \(0\) 
(no interaction). </p>

<p>A network matrix can be created using the function <code>get_network_matrix()</code>. Gene
interaction information for a specific pathway is extracted from the KEGG
database. It is accessed via the function <code>retrieveKGML()</code> from the <code>KEGGgraph</code>
 package. An internet connection is required for this step. </p>

<pre><code>pathw_complete &lt;- get_network_matrix(pathw, directed=FALSE)
</code></pre>

<p>will download the KEGG XML file for the pathway with ID &#39;hsa04710&#39; and save 
it in the working directory. The function will convert the data into a
network matrix and add it to the given pathway object. The expanded pathway 
object will be returned. The user can specify whether the gene-interaction 
matrix should be given directed (directed=TRUE) or undirected (directed=FALSE).</p>

<h2>SNP positions</h2>

<p>Kangar00 offers a function to download positions of the SNPs available in 
your GWAS dataset from the Ensembl database.</p>

<ul>
<li><p><code>snp_info()</code> will take a vector of rs-numbers and give the corresponding base 
pair positions.</p></li>
<li><p>Positions are extracted from the Ensembl database and thus equal the current 
build used on the website. The database is accessed via the function <code>getBM()</code> 
from the package <code>biomaRt</code>. This requires an internet connection. </p></li>
</ul>

<pre><code>snp_info(&quot;rs234&quot;)  
</code></pre>

<p>will return a <code>snp_info</code> object containing the data frame </p>

<table><thead>
<tr>
<th align="center">chr</th>
<th align="center">position</th>
<th align="center">rsnumber</th>
</tr>
</thead><tbody>
<tr>
<td align="center">7</td>
<td align="center">105920689</td>
<td align="center">rs234</td>
</tr>
</tbody></table>

<h2>Pathway Annotation</h2>

<p>To define SNP sets representing a pathway, the function <code>get_anno()</code> can be used. </p>

<ul>
<li><p>Input arguments are a <code>pathway_info</code> as well as a <code>snp_info</code> object.</p></li>
<li><p>If you do not want to change positions in your SNP file using the <code>snp_info()</code>
function, you will have to transform it into a <code>snp_info</code> object
including a data frame listing all SNPs to be annotated. This data frame 
must include the columns &#39;chr&#39;, &#39;position&#39; and &#39;rsnumber&#39;, giving for each SNP 
the chromosome it lies on, its base pair position on the chromosome and the 
rs-numbers identifier, respectively.
See also the output description of <code>snp_info()</code>.</p></li>
<li><p>For annotation the package <code>sqldf</code> is used.</p></li>
</ul>

<pre><code>get_anno(snp_info, pathway_info)
</code></pre>

<p>will return a data frame listing all SNPs that lie inside the boundaries of one 
or more genes in the pathway. That means that genes can appear several times, 
depending on the number of SNPs mapped to them. A SNP can and will be mapped to 
multiple genes if they overlap. The data frame will have the following format</p>

<table><thead>
<tr>
<th align="center">pathway</th>
<th align="right">gene</th>
<th align="right">chr</th>
<th align="right">snp</th>
<th align="right">position</th>
</tr>
</thead><tbody>
<tr>
<td align="center">hsa04710</td>
<td align="right">CSNK1E</td>
<td align="right">22</td>
<td align="right">rs11089885</td>
<td align="right">38413480</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">CSNK1E</td>
<td align="right">22</td>
<td align="right">rs13054361</td>
<td align="right">38336819</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">CSNK1E</td>
<td align="right">22</td>
<td align="right">rs135757</td>
<td align="right">38307648</td>
</tr>
<tr>
<td align="center">hsa04710</td>
<td align="right">&hellip;</td>
<td align="right">&hellip;</td>
<td align="right">&hellip;</td>
<td align="right">&hellip;</td>
</tr>
</tbody></table>

<h2>GWAS data</h2>

<p>Data from a case control study is needed to test a pathways influence on disease
 risk with the logistic kernel machine test in <code>kangar00</code>. Here, GWAS data is 
 represented by the <code>GWASdata</code> object. It includes </p>

<ul>
<li><p>Genotype data for each individual. </p>

<ul>
<li> Genotype data needs to be a matrix with one line per individual and 
  one column for each SNP. </li>
<li> Rownames give ID numbers for the individuals while columnames give the
  rs-numbers corresponding to the SNPs genotyped in the study.<br/></li>
<li> Note that missing values are not allowed and SNPs with missing genotypes
  have to be imputed or excluded from the sample prior to creation of the
 <code>GWASdata</code> object. </li>
</ul></li>
<li><p>Phenotype data for each individual.</p>

<ul>
<li> Phenotypes need to be given in a <code>data frame</code> with the first column 
  including the individual IDs as in the genotype sample.<br/></li>
<li> Further columns can contain informative covariates (such as age, sex,
  &hellip;) to be used in the logistic regression model.<br/></li>
</ul></li>
<li><p>Annotation of study SNPs to pathways created by <code>get_anno</code>. </p>

<ul>
<li> This <code>data frame</code> defines the SNP set representing a specific pathway.
  It can be created using the function <code>get_anno()</code>. </li>
</ul></li>
<li><p>A <code>character</code> describing the data can be added to the <code>GWASdata</code> object. This 
could for instance be the name of the study.    </p></li>
</ul>

<p>A <code>GWASdata</code> object can be constructed as </p>

<pre><code>
my_gwas &lt;- GWASdata(pheno=pheno, geno=geno, anno=anno, desc=&quot;study xy&quot;)


</code></pre>

<h1>Calculation of Kernel Matrices</h1>

<p>Once a <code>GWASdata</code> object is created, we can start to calculate kernel matrices to
test a pathways influence on disease risk. <code>Kangar00</code> offers three different
kernel functions to compute a similarity matrix for the individuals in analysis. 
They will be explained in the following.</p>

<h2>Linear Kernel (Lin)</h2>

<p>The linear kernel assumes additive SNP effects. It is calculated as 
\[ZZ^t         (2)\] 
where \(Z\) denotes the genotype matrix (See also Liu et al, 2010). 
In <code>kangar00</code> a linear kernel can be created using the function <code>kernel_lin()</code>. 
It requires as arguments </p>

<ul>
<li><p>A <code>GWASdata</code> object containing the genotype information.</p></li>
<li><p>A <code>pathway</code> object specifying the pathway to be tested. </p></li>
<li><p>A value for argument <code>calculation</code> to decide how the kernel should be 
calculated. Options are <code>cpu</code> for calculation on cpu and <code>gpu</code>
for gpu calculation.  </p></li>
</ul>

<pre><code>K_lin &lt;- lin_kernel(gwas, p, calculation=&#39;cpu&#39;)
</code></pre>

<p>will return a quadratic <code>matrix</code> of dimension equal to the number of individuals
in the <code>GWASdata</code> object.</p>

<h2>Size-adjusted Kernel (Sia)</h2>

<p>The size-adjusted kernel takes into consideration the numbers of SNPs and genes 
in a pathway to correct for size bias. It is calculated as 
\[ K_{i,j} = exp( - \sqrt{ \frac{1}{r_p} } \sum_{g } ( \frac{|| z_i^g - z_j^g||}
   { \mu_g k_g^{eff}})^{\delta}_g )           (3)\] </p>

<p>Here \(z_i^g\) is the vector of individual i &#39;s genotypes in gene \(g\) and \(r_p\) the
number of genes in pathway \(p\). Scaling parameters \(k_g^{eff}\), \(\mu^g\) and 
\(\delta_g\) adjust for the number of genes in the pathway and the number of SNPs 
within these genes (for more details refer to Freytag et al. 2012).</p>

<p>A kernel of this type can be calculated using the function <code>kernel_sia()</code> with<br/>
the following arguments </p>

<ul>
<li><p>A <code>GWASdata</code> object containing the genotype information.</p></li>
<li><p>A <code>pathway</code> object specifying the pathway to be tested. </p></li>
<li><p>A value for argument <code>calculation</code> to decide how the kernel should be 
calculated. Currently only <code>cpu</code> for cpu calculation is available.  </p></li>
</ul>

<pre><code>K_sia &lt;- sia_kernel(gwas, p, calculation=&#39;cpu&#39;)
</code></pre>

<p>will return a quadratic <code>matrix</code> of dimension equal to the number of individuals
in the <code>GWASdata</code> object.</p>

<h2>Network Kernel (Net)</h2>

<p>The network kernel incorporates information about gene-gene interactions into 
the model. It is defined as 
\[ K = Z A N A^t Z^t      (4)\]<br/>
where matrix \(A\) maps SNPs to genes, \(N\) represents the underlying network 
structure, and \(Z\) is the genotype matrix. 
The network based kernel matrix for a pathway can be calculated with the function
 <code>kernel_net()</code>. Following arguments are needed </p>

<ul>
<li><p>A <code>GWASdata</code> object containing the genotype information.</p></li>
<li><p>A <code>pathway</code> object specifying the pathway to be tested. </p></li>
<li><p>A value for argument &#39;calculation&#39; to decide how the kernel should be 
calculated.   </p></li>
</ul>

<pre><code>K_net &lt;- net_kernel(gwas, p, calculation=&#39;cpu&#39;)
</code></pre>

<p>will return a quadratic <code>matrix</code> of dimension equal to the number of individuals
in the <code>GWASdata</code> object.</p>

<p>Alternatively, kernel matrices can be calculated using the function <code>calc_kernel()</code>.
Here the kernel type is specified via an additional argument <code>type</code>. It can be 
set to lin, sia or net. </p>

<pre><code>K &lt;- calc_kernel(gwas, p, type=&#39;lin&#39;, parallel=&#39;none&#39;)
</code></pre>

<p>This function will simply call the suitable kernel function as described above 
and therefore has the same output. </p>

<h1>Variance Component Test</h1>

<p>A pathways influence on the probability of being a case is evaluated in a 
variance component test. The test statistic is
\[ Q = \frac{1}{2} ( y - \mu)^t K ( y - \mu)  (5)\]
with \(\mu\) the vector of null model estimators given by \(\mu_i = logit^{-1} 
(x_i^t \beta)\) for an individual \(i\) and \(K\) a kernel matrix of the pathway to be
 tested. \(Q\) follows a mixture of \(X^2\) distributions which 
 can be approximated using the Satterthwaite procedure (Schaid 2012) or
 Davies method as implemented in the R package <code>QuadCompForm</code> (Davies 1980). 
  More details on the test can be found in Wu et al (2010). </p>

<p>In <code>kangar00</code> the logistic kernel machine test can be applied to a SNP set 
defining a pathway with the function <code>lkmt</code>. It needs the following arguments</p>

<ul>
<li><p>A formula specifying the null model to be used in the test. The dependent 
 variable is the case control status of the individual (in the example 
 denoted as &#39;pheno&#39;) and is explained by an intercept and optional 
 covariates. </p></li>
<li><p>A linear, size-adjusted or network kernel matrix calculated by one of the
 kernel functions <code>kernel_lin()</code>, <code>kernel_sia()</code> or <code>kernel_net()</code>. </p></li>
<li><p>A  <code>GWASdata</code> object including the genotype based on which the test
 should be performed.</p></li>
<li><p>A <code>character</code> specifying which method should be used to calculate the 
 p-value. Available are &#39;satt&#39; for the Satterthwaite approximation (Schaid
 2010) or &#39;davies&#39; for Davies method (Davies 1980). </p></li>
</ul>

<pre><code>pval_net &lt;- lkmt(pheno ~ 1+sex+age, K_mat, my_gwas, method=&#39;satt&#39;)
</code></pre>

<p>will return an object of type <code>lkmt</code> giving the test result for the pathway 
on which the kernel matrix &#39;K_mat&#39; was calculated. The <code>GWASdata</code> object 
&#39;my_gwas&#39; has to be the same as used to calculate the kernel matrix. 
 The formula above would for example fit for a phenotype file of the
 following format (IDs in first column are always required in phenotype file)</p>

<table><thead>
<tr>
<th align="center">ID</th>
<th align="center">pheno</th>
<th align="center">sex</th>
<th align="center">age</th>
<th align="center">smoker</th>
</tr>
</thead><tbody>
<tr>
<td align="center">ind1</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">41</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">ind2</td>
<td align="center">0</td>
<td align="center">0</td>
<td align="center">38</td>
<td align="center">0</td>
</tr>
<tr>
<td align="center">ind3</td>
<td align="center">1</td>
<td align="center">1</td>
<td align="center">56</td>
<td align="center">1</td>
</tr>
<tr>
<td align="center">&hellip;</td>
<td align="center">&hellip;</td>
<td align="center">&hellip;</td>
<td align="center">&hellip;</td>
<td align="center">&hellip;</td>
</tr>
</tbody></table>

<p>note, that the columns to be used in the model are specified in the formula
given to the <code>lkmt()</code> function and not all covariates have to be used.  </p>

<h1>References</h1>

<ul>
<li><p>Liu D, Ghosh D, Lin X. Estimation and testing for the effect of a genetic 
 pathway on a disease outcome using logistic kernel machine regression via 
 logistic mixed models. BMC Bioinformatics 2008 9:292.</p></li>
<li><p>Schaid DJ: Genomic similarity and kernel methods I: advancements by building
 on mathematical and statistical foundations. Hum Hered 2010, 70:109-131.</p></li>
<li><p>Davies R: Algorithm as 155: the distribution of a linear combination of 
 chi-2 random variables. J R Stat Soc Ser C 1980, 29:323-333.</p></li>
<li><p>Wu MC, Kraft P, Epstein MP, Taylor DM, Chanock SJ, Hunter DJ, Lin X: 
 Powerful SNP-Set Analysis for Case-Control Genome-Wide Association Studies.
 Am J Hum Genet 2010, 86:929-42</p></li>
<li><p>Cunningham F, Amode MR, Barrell D et al. Ensembl 2015. 
 Nucleic Acids Research 2015 43 Database issue:D662-D669</p></li>
<li><p>Kanehisa, M., Goto, S., Sato, Y., Kawashima, M., Furumichi, M., and Tanabe, 
 M.; Data, information, knowledge and principle: back to metabolism in KEGG.
 Nucleic Acids Res. 42, D199-D205 (2014). </p></li>
<li><p>Freytag S, Bickeboeller H, Amos CI, Kneib T, Schlather M: A Novel Kernel 
 for Correcting Size Bias in the Logistic Kernel Machine Test with an 
 Application to Rheumatoid Arthritis. Hum Hered. 2012, 74(2):97-108.</p></li>
<li><p>Freytag S, Manitz J, Schlather M, Kneib T, Amos CI, Risch A, Chang-Claude J,
 Heinrich J, Bickeboeller H: A network-based kernel machine test for the 
 identification of risk pathways in genome-wide association studies. Hum 
 Hered. 2013, 76(2):64-75.</p></li>
</ul>

</body>

</html>
